lock-api-objects/src/main/java/com/palantir/lock/watch/LockMetadata.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * (c) Copyright 2023 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.watch;
+
+import com.fasterxml.jackson.annotation.JsonSubTypes;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.fasterxml.jackson.annotation.JsonTypeName;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import com.palantir.lock.watch.LockMetadata.Created;
+import com.palantir.lock.watch.LockMetadata.Deleted;
+import com.palantir.lock.watch.LockMetadata.Unchanged;
+import com.palantir.lock.watch.LockMetadata.Updated;
+import org.immutables.value.Value;
+import org.immutables.value.Value.Immutable;
+
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
+@JsonSubTypes({
+    @JsonSubTypes.Type(value = Unchanged.class, name = Unchanged.TYPE),
+    @JsonSubTypes.Type(value = Updated.class, name = Updated.TYPE),
+    @JsonSubTypes.Type(value = Deleted.class, name = Deleted.TYPE),
+    @JsonSubTypes.Type(value = Created.class, name = Created.TYPE),
+})
+public interface LockMetadata {
+    <T> T accept(Visitor<T> visitor);
+
+    static Unchanged unchanged() {
+        return ImmutableUnchanged.builder().build();
+    }
+
+    static Updated update(byte[] oldValue, byte[] newValue) {
+        return ImmutableUpdated.of(oldValue, newValue);
+    }
+
+    static Deleted delete(byte[] deletedValue) {
+        return ImmutableDeleted.of(deletedValue);
+    }
+
+    static Created create(byte[] createdValue) {
+        return ImmutableCreated.of(createdValue);
+    }
+
+    interface Visitor<T> {
+        T visit(Unchanged unchanged);
+
+        T visit(Updated updated);
+
+        T visit(Deleted deleted);
+
+        T visit(Created created);
+    }
+
+    /**
+     * LockMetadata event indicating that a value has definitely not been changed.
+     */
+    @Immutable
+    @JsonSerialize(as = ImmutableUnchanged.class)
+    @JsonDeserialize(as = ImmutableUnchanged.class)
+    @JsonTypeName(Unchanged.TYPE)
+    interface Unchanged extends LockMetadata {
```

Comment:
Metadata subtypes will be accessed like so: `LockMetadata.Unchanged`

Reply:
I don't think we _need_ the value here since, if nothing changed, anything we have cached is still valid. As mentioned below, we might want to include it anyways for validation purposes. I'd expect to run into the `Unchanged` case fairly often so adding the value here does introduce significant overhead on the wire.